ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+      if (firstChunk.isComplete() && readCalled) {
+        while (operationException.get() == null && chunkIndexToBuffer.containsKey(indexOfNextChunkToWriteOut)) {
+          ByteBuffer chunkBuf = chunkIndexToBuffer.remove(indexOfNextChunkToWriteOut);
+          asyncWritableChannel.write(chunkBuf, chunkAsyncWriteCallback);
+          indexOfNextChunkToWriteOut++;
+        }
+        if (operationException.get() != null || numChunksWrittenOut.get() == numChunksTotal) {
+          completeRead();
+        }
+      }
+    }
+
+    /**
+     * Complete the read from this {@link ReadableStreamChannel} by invoking the callback and marking the future.
+     */
+    void completeRead() {
+      if (readIntoCallbackCalled.compareAndSet(false, true)) {
+        streamSize = bytesWritten;
+        readIntoFuture.done(bytesWritten, operationException.get());
+        if (readIntoCallback != null) {
+          readIntoCallback.onCompletion(bytesWritten, operationException.get());
+        }
+      }
+      operationCompleted = true;
+    }
+  }
+
+  /**
+   * GetChunk is used to retrieve and hold a data chunk of a composite blob. An object of this class is initialized
+   * to retrieve one data chunk at a time. Once the associated chunk is successfully retrieved, this object can be
+   * reinitialized and used to retrieve a subsequent chunk.
+   */
+  private class GetChunk {
+    // the operation tracker used to track the operation on the current chunk.
+    private OperationTracker chunkOperationTracker;
+    // whether the operation on the current chunk has completed.
+    private boolean chunkCompleted;
+    // the blob id of the current chunk.
+    private BlobId chunkBlobId;
+    // the index of the current chunk in the overall blob.
+    private int chunkIndex;
+    // the most relevant exception encountered for the current chunk.
+    protected RouterException chunkException;
+    // For a GetChunk, responses may be handled multiple times. Regardless of the successTarget,
+    // the actual body of the response is deserialized only once.
+    protected boolean successfullyDeserialized;
+    // map of correlation id to the request metadata for every request issued for this operation.
+    protected final Map<Integer, GetRequestInfo> correlationIdToGetRequestInfo = new TreeMap<>();
+    // the state of the chunk.
+    protected volatile ChunkState state;
+
+    /**
+     * Construct a GetChunk
+     * @param index the index (in the overall blob) of the initial data chunk that this GetChunk has to fetch.
+     * @param id the {@link BlobId} of the initial data chunk that this GetChunk has to fetch.
+     */
+    GetChunk(int index, BlobId id) {
+      reset();
+      initialize(index, id);
+    }
+
+    /**
+     * Do what needs to be done after the GetBlobOperation is done with the current chunk that has completed. When
+     * this is called, the state of the chunk should be {@link ChunkState#Complete}
+     */
+    void postCompletionCleanup() {
```

Comment:
`FirstChunk` is also a `GetChunk` and `reset()` is part of `GetChunk` constructor.